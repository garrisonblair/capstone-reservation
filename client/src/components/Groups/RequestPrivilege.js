import React, { Component } from 'react';
import { Button, Dropdown, Label } from 'semantic-ui-react';
import PropTypes from 'prop-types';
import sweetAlert from 'sweetalert2';
import api from '../../utils/api';
import './RequestPrivilege.scss';

class RequestPrivilege extends Component {
  state = {
    options: [],
    dropdownValue: '',
    labelText: 'none',
    labelColor: 'grey',
    buttonText: 'Request Privilege',
    currentStatus: 'new',
    disableDropdown: false,
    groupId: null,
    groupPrivilege: null,
  }

  componentDidMount() {
    const { groupId, groupPrivilege, justForTesting } = this.props;
    // This is just for testing.
    if (justForTesting) {
      const { options } = this.state;
      options.push({ value: 1, text: 'option 1', key: 1 });
    }

    this.setState({ groupId, groupPrivilege });
    api.getPrivileges()
      .then((r) => {
        if (r.status === 200) {
          const { options } = this.state;
          r.data.filter(p => p.is_default === false)
            .map(p => options.push({ value: p.id, text: p.name, key: p.id }));
          this.setState({ options });
        } else {
          sweetAlert(`${r.status} error`, r.text, 'error');
        }
      })
      .catch((r) => {
        console.log(r);
        sweetAlert('Something went wrong', r.text, 'error');
      });
    if (groupPrivilege !== null) {
      this.setState({ dropdownValue: groupPrivilege.privilege_category });
      if (groupPrivilege.status === 'Pending') {
        this.changeToPending();
      } else if (groupPrivilege.status === 'Denied') {
        this.changeToDenied();
      } else if (groupPrivilege.status === 'Approved') {
        this.changeToApproved();
      }
    }
  }

  handleRequestPrivilege = () => {
    const { dropdownValue, groupId } = this.state;
    if (dropdownValue.length === 0) {
      sweetAlert('Empty field', 'Please choose a privilege', 'warning');
      return;
    }
    api.requestPrivilege(groupId, dropdownValue)
      .then((r) => {
        if (r.status === 201) {
          this.changeToPending();
          this.setState({ groupPrivilege: r.data });
        } else {
          sweetAlert(`${r.status} error`, r.text, 'error');
        }
      })
      .catch((r) => {
        console.log(r);
        sweetAlert('Something went wrong', r.text, 'error');
      });
  }

  handleCancelRequest = () => {
    const { groupPrivilege } = this.state;
    api.cancelPrivilegeRequest(groupPrivilege.id)
      .then((r) => {
        if (r.status === 200) {
          this.changeToRequest();
        } else {
          sweetAlert(`${r.status} error`, r.text, 'error');
        }
      })
      .catch((r) => {
        console.log(r);
        sweetAlert('Something went wrong', r.text, 'error');
      });
  }

  handlePrivilegeChange = (e, v) => {
    this.setState({ dropdownValue: v.value });
  }

  changeToRequest = () => {
    this.setState({
      labelColor: 'grey',
      buttonText: 'Request Privilege',
      labelText: 'none',
      disableDropdown: false,
      currentStatus: 'new',
    });
  }

  changeToPending = () => {
    this.setState({
      labelColor: 'yellow',
      buttonText: 'Cancel Request',
      labelText: 'pending',
      disableDropdown: true,
      currentStatus: 'pending',
    });
  }

  changeToApproved = () => {
    this.setState({
      labelColor: 'green',
      buttonText: '',
      labelText: 'approved',
      disableDropdown: true,
      currentStatus: 'approved',
    });
  }

  changeToDenied = () => {
    this.setState({
      labelColor: 'red',
      buttonText: 'Make new Request',
      labelText: 'denied',
      disableDropdown: true,
      currentStatus: 'denied',
    });
  }

  buttonOnClick = () => {
    const { currentStatus } = this.state;
    if (currentStatus === 'new') {
      this.handleRequestPrivilege();
    } else if (currentStatus === 'pending') {
      this.handleCancelRequest();
    } else if (currentStatus === 'denied') {
      this.changeToRequest();
    }
  }

  render() {
    const {
      options, labelText, labelColor, buttonText, disableDropdown, dropdownValue, currentStatus,
    } = this.state;
    // if there is no privilege show nothing.
    if (options.length === 0) {
      return ('');
    }
    return (
      <div id="groups-request-privilege">
        <div>
          <h3>Privilege</h3>
          <Label color={labelColor}>{labelText}</Label>
        </div>
        <br />
        <Dropdown
          disabled={disableDropdown}
          placeholder="Privileges"
          options={options}
          selection
          onChange={this.handlePrivilegeChange}
          value={dropdownValue}
        />
        {currentStatus !== 'approved' ? (
          <Button
            color="blue"
            onClick={this.buttonOnClick}
            className="button-right"
          >
            {buttonText}
          </Button>
        ) : null}
      </div>
    );
  }
}

RequestPrivilege.propTypes = {
  groupId: PropTypes.number.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  groupPrivilege: PropTypes.object,
  justForTesting: PropTypes.bool,
};

RequestPrivilege.defaultProps = {
  groupPrivilege: null,
  justForTesting: false,
};

export default RequestPrivilege;
