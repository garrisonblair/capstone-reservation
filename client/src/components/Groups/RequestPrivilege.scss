#groups-request-privilege{
    h3{
        float: left;
    }
    .button-right{
        margin-left: 10px;
    }
}
