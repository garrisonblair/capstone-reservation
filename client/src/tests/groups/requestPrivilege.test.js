import React from 'react';
import renderer from 'react-test-renderer';
import RequestPrivilege from '../../components/Groups/RequestPrivilege';


it('Shows empty RequestPrivilege', () => {
  const tree = renderer
    .create(
      <RequestPrivilege
        groupId={1}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Shows new RequestPrivilege', () => {
  const tree = renderer
    .create(
      <RequestPrivilege
        groupId={1}
        justForTesting
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
