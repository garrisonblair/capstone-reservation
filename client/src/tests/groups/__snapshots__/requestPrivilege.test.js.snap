// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shows empty RequestPrivilege 1`] = `""`;

exports[`Shows new RequestPrivilege 1`] = `
<div
  id="groups-request-privilege"
>
  <div>
    <h3>
      Privilege
    </h3>
    <div
      className="ui grey label"
      onClick={[Function]}
    >
      none
    </div>
  </div>
  <br />
  <div
    aria-disabled={false}
    aria-expanded={false}
    className="ui selection dropdown"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    role="listbox"
    tabIndex={0}
  >
    <div
      aria-live="polite"
      className="default text"
      role="alert"
    >
      Privileges
    </div>
    <i
      aria-hidden="true"
      className="dropdown icon"
      onClick={[Function]}
    />
    <div
      className="menu transition"
    >
      <div
        aria-checked={false}
        aria-selected={true}
        className="selected item"
        onClick={[Function]}
        role="option"
        style={
          Object {
            "pointerEvents": "all",
          }
        }
      >
        <span
          className="text"
        >
          option 1
        </span>
      </div>
    </div>
  </div>
  <button
    className="ui blue button button-right"
    onClick={[Function]}
  >
    Request Privilege
  </button>
</div>
`;
